Frontend/bugfix/show on map

thdk opened the pull request:
Closes #1786 

Fixes/changes:
- On Search page, when you click or tap the 'show user on map' button, it should always center the map on that user. Even if that user was already selected.
- On Search page, when you click or tap the marker of the current selected user on the map, that marker should NOT be deselected. (to deselect, user can click any empty space on the map)

**Frontend checklist**
- [x] Formatted my code with `yarn format && yarn lint --fix`
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [ ] Added any new components to storybook
- [ ] Added tests where relevant
- [x] All tests pass (except for the one listed below)

```shell
 FAIL  src/components/Datepicker.test.tsx (5.272 s)
  ● DatePicker › should submit with proper date for clicking

    expect(received).toEqual(expected) // deep equality

    Expected: "2021-03-23T00:00:00+01:00"
    Received: "2021-03-23T01:00:00+01:00"
```
- [x] Clicked around my changes running locally and it works
- [x] Checked Desktop, Mobile and Tablet screen sizes

thdk committed:
fix map click event handlers being called twice

When an active feature on the map was clicked, both the
`handleMapUserClick` and `handleMapClickAway` callbacks where called.

Reason for this was the order in which these event callbacks are
initiated. `handleMapUserClick` calls prevent default on the event object
and `handleMapClickAway` checks if `event.preventDefault()` was called
already. So handleMapUserClick should be bound before
handleMapClickAway.

Also removed an (imho) unneeded dependency of the useEffect hook.

thdk committed:
fix show user on map when user is already selected

thdk committed:
linted all code for frontend

darrenvong commented on app/frontend/src/features/search/SearchPage.tsx:
```diff
@@ -159,28 +162,32 @@ export default function SearchPage() {
     };
 
-    map.current.on("click", handleMapClickAway);
+    //bind event handlers for map events (order matters!)
     map.current.on(
       "click",
       layers.unclusteredPointLayer.id,
       handleMapUserClick
     );
+    map.current.on("click", handleMapClickAway);
```
I'm not sure I completely following this... what was the behaviour of the old ordering? It wrongly deselects the user on the map since it calls `handleResultClick` with `undefined`?

Whereas now, I'm guessing the `ev.preventDefault` of `handleMapUserClick` would stop it from being called with this new ordering?

thdk replied:
Steps to reproduce the issue:  (old behavior)
- go to search page
- search for a user
- click 'show user on map button
- the user's map marker is now active on the map
- click this active marker on the map

**expected:** the user's marker should stay active and the map should pan so the marker is centered on the map again if it wasn't centered anymore.
**actual:** the user's marker is deselected

thdk replied:
without this change, the `handleResultClick` of `SearchPage.tsx` was always called twice when you click a marker on the map. The first time (handleMapClickAway) the value of `user` argument was always undefined. The second time (handleMapUerClick) the value of `user` argument is a User object.

You can test this in production when you add a breakpoint on the first line of `handleResultClick` method.

darrenvong reviewed (commented):
In the commit message, did you mean to say:

```
So handleMapUserClick should be bound before handleMapClickAway.
```

Would be nice to correct it since it's a good commit message with context!

thdk committed:
format all frontend code

thdk commented:
Rebased my code with two changes:
- fixed the commit message in one of the commits
- removed the empty file SearchPage.test.tsx which I wrongly committed together with the formatting commit

thdk commented:
Preview url: https://c803b369--frontend.preview.couchershq.org/search?query=an#loc=2.48/35.68/139.76

darrenvong reviewed (approved):
I thought the fix would be more involved than this. Some nice detective work there! 👍
